compiler/rustc_resolve/src/diagnostics.rs

```diff
@@ -1421,14 +1419,42 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
             "",
         );
 
+        if macro_kind == MacroKind::Bang
+            && ident.name == sym::macro_rules
+            && let source_map = self.tcx.sess.source_map()
+            && let Ok(next_source) = source_map.span_to_next_source(ident.span)
+            && let Some(brace_offset) = {
+                // find the first `{`
+                let chars = next_source.chars().collect::<Vec<_>>();
+                if chars.len() < 2 || chars[0] != '!' {
+                    None
+                } else {
+                    let mut idx = 1;
+                    while idx < chars.len() && chars[idx].is_ascii_whitespace() {
+                        idx += 1;
+                    }
+                    (idx < chars.len() && chars[idx] == '{').then_some(idx)
+                }
+            }
+        {
+            let bang_pos = ident.span.hi().0;
+            let lo = bang_pos + 1; // the pos after `!`
+            let hi = bang_pos + brace_offset as u32; // the pos of '{'
```

Comment:
It's safe because we have ensured the inclusion of `﻿!` and `{`

Reply:
This is not obvious, and the logic is now overcomplicated too much.
One single call to `span_extend_while` (or a similar method) to find where `!` ends is the maximum amount of logic that I'd want to see here from the cost/benefit point of view.

Reply:
I think the current method simplifies the process while still preserving the tip